### PR TITLE
Add ADJouleHeatingSource and ADElectricalConductivity

### LINKED ISSUE #15536: Add ADJouleHeatingSource
## Reason
`ADJouleHeatingSource` is needed in order to properly handle cases when a Joule heating source is needed and an AD electrical conductivity is required.

## Design
Create an AD version of `JouleHeatingSource`

## Impact
New capability.

### PR DESCRIPTION
**Pull request by cticenhour**
In order to completely AD-ify the test I created for `ADJouleHeatingSource`, I went ahead and created a templated version of `ADElectricalConductivity`. 

Closes #15536

### CODE REVIEW

**Comment by cticenhour** [comment]
Tagging @dschwen for review

File: modules/heat_conduction/include/kernels/ADJouleHeatingSource.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ */
+class ADJouleHeatingSource : public ADKernel
+{
+public:
+  static InputParameters validParams();
+
+  ADJouleHeatingSource(const InputParameters & parameters);
+
+protected:
+  virtual ADReal computeQpResidual();
```
Comment (lindsayad) [inline modules/heat_conduction/include/kernels/ADJouleHeatingSource.h:26]: mark `override`

File: modules/heat_conduction/test/tests/joule_heating/transient_ad_jouleheating.i [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    full = true
+  [../]
+[]
+
+[Executioner]
+  type = Transient
+  scheme = bdf2
+  solve_type = NEWTON
+  petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
+  petsc_options_value = 'asm         101   preonly   ilu      1'
```
Comment (lindsayad) [inline modules/heat_conduction/test/tests/joule_heating/transient_ad_jouleheating.i:94]: I'm curious why `preonly`? Also no need to remove the explicit setting of `-pc_asm_overlap`, but the default asm overlap is 1.

File: modules/heat_conduction/src/kernels/ADJouleHeatingSource.C [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  : ADKernel(parameters),
+    _grad_elec(adCoupledGradient("elec")),
+    _elec_cond(getADMaterialProperty<Real>("electrical_conductivity"))
+{
+}
+
+ADReal
+ADJouleHeatingSource::computeQpResidual()
+{
+  return -_elec_cond[_qp] * _grad_elec[_qp] * _grad_elec[_qp] * _test[_i][_qp];
```
Comment (lindsayad) [inline modules/heat_conduction/src/kernels/ADJouleHeatingSource.C:39]: I'd recommend inheriting from `ADKernelValue` for performance

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: modules/heat_conduction/test/tests/joule_heating/transient_ad_jouleheating.i (hunk 2 above)
Comment (cticenhour) [reply modules/heat_conduction/test/tests/joule_heating/transient_ad_jouleheating.i:94]: I pulled this from the previous `transient_jouleheating.i` test, which had `type = PJFNK` so I am unfamiliar with the specific set points used here. Is `preonly` commonly only used with PJFNK?

**Comment by lindsayad** [comment]
I've only ever used it myself when doing NEWTON with `-pc_type lu` since I
know that it should solve in a single iteration without a need to apply
GMRES

**Comment by cticenhour** [comment]
I've done the same. I'll adjust these settings in the next revision.

**Comment by cticenhour** [comment]
I saw the best convergence with
```
  petsc_options_iname = '-pc_type -sub_ksp_type'
  petsc_options_value = 'lu       preonly'
```
So I stuck with those settings. I also addressed your other comments.

**Comment by lindsayad** [comment]
How does it do with just the options: `-pc_type hypre`?

**Comment by cticenhour** [comment]
With `hypre` for time step 5:
```
Time Step 5, time = 5, dt = 1
 0 Nonlinear |R| = 1.910580e+01
      0 Linear |R| = 1.910580e+01
      1 Linear |R| = 1.023313e-01
      2 Linear |R| = 1.882281e-03
 1 Nonlinear |R| = 1.887918e-03
      0 Linear |R| = 1.887918e-03
      1 Linear |R| = 8.455513e-05
      2 Linear |R| = 3.962629e-06
      3 Linear |R| = 1.402988e-07
 2 Nonlinear |R| = 1.402004e-07
 Solve Converged!
```

With `lu` at the same time step
```
Time Step 5, time = 5, dt = 1
 0 Nonlinear |R| = 1.910580e+01
      0 Linear |R| = 1.910580e+01
      1 Linear |R| = 5.043881e-15
 1 Nonlinear |R| = 8.508753e-05
      0 Linear |R| = 8.508753e-05
      1 Linear |R| = 2.841526e-20
 2 Nonlinear |R| = 5.723515e-12
 Solve Converged!
```

This timestep is representative of the rest of the solve.

**Comment by lindsayad** [comment]
That looks like a really efficient linear solve to me with `hypre`. Let's go with that. And go ahead and remove the `nl_abs_tol` since we try to discourage use of it. Does the problem approach a steady-state over time?

**Comment by lindsayad** [comment]
I guess that's the last time step so I can not be lazy and answer my own question :smile: Yea let's remove `nl_abs_tol`

**Comment by cticenhour** [comment]
Yeah - that's the last one 😄  I'll remove that. I'll also remove `l_tol`, since that's also another old holdover from prior to AD and automatic scaling. Convergence removing all tolerances looks like:

```
Time Step 5, time = 5, dt = 1
 0 Nonlinear |R| = 1.910580e+01
      0 Linear |R| = 1.910580e+01
      1 Linear |R| = 1.023313e-01
      2 Linear |R| = 1.882281e-03
      3 Linear |R| = 7.508153e-05
 1 Nonlinear |R| = 1.122902e-04
      0 Linear |R| = 1.122902e-04
      1 Linear |R| = 6.308807e-06
      2 Linear |R| = 1.300444e-07
      3 Linear |R| = 4.332178e-09
      4 Linear |R| = 1.763024e-10
 2 Nonlinear |R| = 1.771571e-10
 Solve Converged!
```

**Comment by lindsayad** [comment]
Looks superb

**Comment by cticenhour** [comment]
Alright, I think we're good to go then, pending some random testing failure.

**Review by lindsayad — APPROVED** [review]

**Pull request merged**